src/elm/external.ts

```diff
@@ -22,21 +25,49 @@ export class Retrieve extends Expression {
   }
 
   exec(ctx: Context) {
-    let records = ctx.findRecords(this.templateId != null ? this.templateId : this.datatype, this);
-    let codes = this.codes;
-    if (this.codes && typeof this.codes.exec === 'function') {
-      codes = this.codes.execute(ctx);
-      if (codes == null) {
+    // Object with retrieve information to pass back to patient source
+    const retrieveDetails: RetrieveDetails = {
+      datatype: this.datatype
+    };
+
+    let resolvedCodes: Code[] | ValueSet | undefined;
+    if (this.codes) {
+      resolvedCodes = this.codes.execute(ctx);
+
+      if (resolvedCodes == null) {
         return [];
       }
-    }
-    if (codes) {
-      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, codes));
+
+      retrieveDetails.codes = resolvedCodes;
     }
     // TODO: Added @dateProperty check due to previous fix in cql4browsers in cql_qdm_patient_api hash: ddbc57
+    let range: Interval | undefined;
     if (this.dateRange && this.dateProperty) {
-      const range = this.dateRange.execute(ctx);
-      records = records.filter((r: any) => range.includes(r.getDateOrInterval(this.dateProperty)));
+      range = this.dateRange.execute(ctx);
+
+      retrieveDetails.dateRange = range;
+      retrieveDetails.dateProperty = this.dateProperty;
+    }
+
+    if (this.templateId) {
+      retrieveDetails.templateId = this.templateId;
+    }
+
+    if (resolvedCodes && this.codeProperty) {
+      retrieveDetails.codeProperty = this.codeProperty;
+    }
+
+    let records = ctx.findRecords(
+      this.templateId != null ? this.templateId : this.datatype,
+      retrieveDetails
+    );
+
+    if (resolvedCodes) {
+      records = records.filter((r: any) => this.recordMatchesCodesOrVS(r, resolvedCodes));
+    }
+
+    if (range && this.dateProperty) {
```

Comment:
Similarly to `resolvedCodes` above, if you used `retrieveDetails.range` then you could take `range` out of this scope and just keep it in the smaller scope as a `const`.